src/combinations.rs

```diff
@@ -77,6 +77,21 @@ impl<I: Iterator> Combinations<I> {
             self.pool.prefill(k);
         }
     }
+
+    /// For a given size `n`, return the count of remaining elements or None if it would overflow.
+    fn remaining_for(&self, n: usize) -> Option<usize> {
+        let k = self.k();
+        if self.first {
+            checked_binomial(n, k)
+        } else {
+            self.indices
+                .iter()
+                .enumerate()
+                .fold(Some(0), |sum, (k0, n0)| {
+                    sum.and_then(|s| s.checked_add(checked_binomial(n - 1 - *n0, k - k0)?))
+                })
```

Comment:
But I'm pretty sure I *don't trivially* understand this one. Can you shed some light on this or can anyone co-review?

Reply:
Well **this is definitely the hard part**, I'm not gonna say it's trivial, even for me.
The `indices` we track are basically the the k-combinations of `0..n`.
I worked on an example, let's say k=3 and n=5 (big enough to see things properly without having too many elements).
The indices are then (one line per combination):

    0 1 2
    0 1 3
    0 1 4
    0 2 3
    0 2 4
    0 3 4
    1 2 3
    1 2 4
    1 3 4
    2 3 4

There are `binomial(5, 3)==10` elements: **6** elements starting with 0 (`binomial(4,2)` to choose 2 elements from `1..5`), **3** with 1 (`binomial(3,2)` to choose 2 elements from `2..5`), **1** with 2 (`binomial(2,2)` to choose 2 elements from `3..5`).

The binomials (rows for n, columns for k, both starting at 0).

    1
    1 1
    1 2 *1*
    1 3 *3* 1
    1 4 *6* 4 1
    1 5 10 *10* 5 1

Now, let's say we have `indices = 0 1 4` and we want to count the remaining ones. So
- the only one starting with 2 `2 3 4`
- the ones starting with 1 (`1 2 3, 1 2 4, 1 3 4` start with 1 and the rest: choose 2 from 2..5 so `binomial(5-2, 2)`)
- and then for the ones starting with 0, some other binomial**s**:
  - the ones for 3 in 2nd position: `0 3 4` (`binomial(1, 1)` as the only choice is 4)
  - the ones for 2 in 2nd position: `0 2 3, 0 2 4` (`binomial(2, 1)` to choose 1 elements from `3..5`)

So we have a sum of four binomials while we sum `k=3` binomials in the code so you could be concerned. But there is a trick (binomials are fun). See the above triangle where the column 1 3 6 sums into 10, that's a property of the binomials that simplifies the sum into the one in the code. For the example above: `binomial(1,1)+binomial(2,1)==binomial(3,2)`

That's some light, not some hard proof I would give to a math teacher but this is why I put a test where I check the count and the size hint at every step of it.
**PS:** I hope I did not make some mistake in this big comment...